server/src/main/java/org/elasticsearch/index/codec/vectors/es818/OptimizedScalarQuantizer.java

```diff
@@ -0,0 +1,218 @@
+                sumQuery += assignment;
+                destinations[i][h] = (byte) assignment;
+            }
+            results[i] = new QuantizationResult(
+                intervalScratch[0],
+                intervalScratch[1],
+                similarityFunction == EUCLIDEAN ? norm2 : centroidDot,
+                sumQuery
+            );
+        }
+        return results;
+    }
+
+    public QuantizationResult scalarQuantize(float[] vector, byte[] destination, byte bits, float[] centroid) {
+        assert similarityFunction != COSINE || VectorUtil.isUnitVector(vector);
+        assert similarityFunction != COSINE || VectorUtil.isUnitVector(centroid);
+        assert vector.length <= destination.length;
+        assert bits > 0 && bits <= 8;
+        float[] intervalScratch = new float[2];
+        int points = 1 << bits;
+        double vecMean = 0;
+        double vecVar = 0;
+        float norm2 = 0;
+        float centroidDot = 0;
+        float min = Float.MAX_VALUE;
+        float max = -Float.MAX_VALUE;
+        for (int i = 0; i < vector.length; ++i) {
+            if (similarityFunction != EUCLIDEAN) {
+                centroidDot += vector[i] * centroid[i];
+            }
+            vector[i] = vector[i] - centroid[i];
+            min = Math.min(min, vector[i]);
+            max = Math.max(max, vector[i]);
+            norm2 += (vector[i] * vector[i]);
+            double delta = vector[i] - vecMean;
+            vecMean += delta / (i + 1);
+            vecVar += delta * (vector[i] - vecMean);
+        }
+        vecVar /= vector.length;
+        double vecStd = Math.sqrt(vecVar);
+        intervalScratch[0] = (float) clamp((MINIMUM_MSE_GRID[bits - 1][0] + vecMean) * vecStd, min, max);
+        intervalScratch[1] = (float) clamp((MINIMUM_MSE_GRID[bits - 1][1] + vecMean) * vecStd, min, max);
+        optimizeIntervals(intervalScratch, vector, norm2, points);
+        float nSteps = ((1 << bits) - 1);
+        float a = intervalScratch[0];
+        float b = intervalScratch[1];
+        float step = (b - a) / nSteps;
+        int sumQuery = 0;
+        for (int h = 0; h < vector.length; h++) {
+            float xi = (float) clamp(vector[h], a, b);
+            int assignment = Math.round((xi - a) / step);
+            sumQuery += assignment;
+            destination[h] = (byte) assignment;
+        }
+        return new QuantizationResult(
+            intervalScratch[0],
+            intervalScratch[1],
+            similarityFunction == EUCLIDEAN ? norm2 : centroidDot,
+            sumQuery
+        );
+    }
+
+    private double loss(float[] vector, float[] interval, int points, float norm2) {
+        double a = interval[0];
+        double b = interval[1];
+        double step = ((b - a) / (points - 1.0F));
+        double stepInv = 1.0 / step;
+        double xe = 0.0;
+        double e = 0.0;
+        for (double xi : vector) {
+            double xiq = (a + step * Math.round((clamp(xi, a, b) - a) * stepInv));
```

Comment:
yeap, this seems about right 😅

Reply:
LOL, I should indeed add some comments here. We are basically calculating the error of quantizing and then unquantizing and shifting up or down based on the mis-calculation.